src/graphql/graphql-api.php

```diff
@@ -0,0 +1,255 @@
+<?php
+/**
+ * GraphQL API
+ *
+ * @package vip-block-data-api
+ */
+
+namespace WPCOMVIP\BlockDataApi;
+
+use GraphQLRelay\Relay;
+
+defined( 'ABSPATH' ) || die();
+
+/**
+ * GraphQL API to offer an alternative to the REST API.
+ */
+class GraphQLApi {
+	/**
+	 * Initiatilize the graphQL API by hooking into the graphql_register_types action, 
+	 * which only fires if WPGraphQL is installed and enabled, and is further controlled 
+	 * by the vip_block_data_api__is_graphql_enabled filter.
+	 */
+	public static function init() {
+		add_action( 'graphql_register_types', [ __CLASS__, 'register_types' ] );
+	}
+
+	/**
+	 * Extract the blocks data for a post, and return back in the format expected by the graphQL API.
+	 *
+	 * @param  \WPGraphQL\Model\Post $post_model Post model for post.
+	 * 
+	 * @return array
+	 */
+	public static function get_blocks_data( $post_model ) {
+		$post_id = $post_model->ID;
+		$post    = get_post( $post_id );
+
+		$content_parser = new ContentParser();
+
+		$parser_results = $content_parser->parse( $post->post_content, $post_id );
+
+		// We need to not return a WP_Error object, and so a regular exception is returned.
+		if ( is_wp_error( $parser_results ) ) {
+			Analytics::record_error( $parser_results );
+
+			// Return API-safe error with extra data (e.g. stack trace) removed.
+			return new \Exception( $parser_results->get_error_message() );
+		}
+
+		$parser_results['blocks'] = array_map(function ( $block ) {
+			return self::transform_block_format( $block );
+		}, $parser_results['blocks'] );
+
+		return $parser_results;
+	}
+
+	/**
+	 * Transform the block's format to the format expected by the graphQL API.
+	 *
+	 * @param array $block An associative array of parsed block data with keys 'name' and 'attribute'.
+	 *
+	 * @return array
+	 */
+	public static function transform_block_format( $block ) {
+		// Generate a unique ID for the block.
+		$block['id'] = Relay::toGlobalId( 'ID', wp_unique_id() );
+
+		// Convert the attributes to be in the name-value format that the schema expects.
+		$block = self::map_attributes( $block );
+
+		if ( isset( $block['innerBlocks'] ) && ! empty( $block['innerBlocks'] ) ) {
+			$block['innerBlocks'] = array_map( function ( $block ) {
+				return self::transform_block_format( $block );
+			}, $block['innerBlocks'] );
+		}
+
+		return $block;
+	}
+
+	/**
+	 * Convert the attributes to be in the name-value format that the schema expects.
+	 *
+	 * @param array $block An associative array of parsed block data with keys 'name' and 'attribute'.
+	 *
+	 * @return array
+	 */
+	public static function map_attributes( $block ) {
+		// check if type of attributes is stdClass and unset it as that's not supported by graphQL.
+		if ( isset( $block['attributes'] ) && is_object( $block['attributes'] ) ) {
+			unset( $block['attributes'] );
```

Comment:
how / why would this happen?

Reply:
@ingeniumed Can you explain a bit more about the `stdClass` comment here? I'm not sure where that comes from either.

Reply:
So this is because in the contentParser we set the actual attributes to be `(object)[]` which allows it to be given back as empty json in the rest api rather than as an empty array. Due to that, the internal data type comes back as stdClass rather than an array. This workaround is due to that.

I experimented with switching the `(object)[]` to a new ArrayObject but in that case it adds an empty array with key `storage` under the attributes which requires an even worse workaround. That's why it's been done this way.

I tried to search for a better solution but couldn't quite find one and I didn't want to break the rest response or the way the public function works because there are customers already using it.

[This](https://github.com/Automattic/vip-block-data-api/blob/trunk/src/parser/content-parser.php#L268) is the line I'm referencing btw in the contentParser.

Reply:
Hopefully that explains it properly

Reply:
@ingeniumed That sounds familiar now! As a side-effect of returning data in JSON via the REST API, if we want to return `attributes: {}` in a response, we need to cast `[]` into an object to avoid returning an array instead. This is just dealing with that unusual data in the GraphQL layer.